

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>PopGenStatistics Documentation &mdash; SNPio 1.2.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=859da57e" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=590429e0"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Changelog" href="changelog.html" />
    <link rel="prev" title="Example script" href="example_script.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SNPio
              <img src="_static/snpio_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_script.html">Example script</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">PopGenStatistics Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#PopGenStatistics"><code class="docutils literal notranslate"><span class="pre">PopGenStatistics</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#key-features">Key Features</a></li>
<li class="toctree-l2"><a class="reference internal" href="#dependencies">Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quick-start">Quick Start</a></li>
<li class="toctree-l2"><a class="reference internal" href="#methods-overview">Methods Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#calculated-statistics">Calculated Statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="#advanced-usage-and-best-practices">Advanced Usage and Best Practices</a></li>
<li class="toctree-l2"><a class="reference internal" href="#additional-information">Additional Information</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">snpio</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SNPio</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">PopGenStatistics Documentation</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/btmartin721/SNPio/blob/master/docs/source/pop_gen_statistics.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="popgenstatistics-documentation">
<h1>PopGenStatistics Documentation<a class="headerlink" href="#popgenstatistics-documentation" title="Link to this heading"></a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="PopGenStatistics">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">PopGenStatistics</span></span><a class="headerlink" href="#PopGenStatistics" title="Link to this definition"></a></dt>
<dd></dd></dl>

<p>The <code class="docutils literal notranslate"><span class="pre">PopGenStatistics</span></code> class is designed to perform a suite of population genetic analyses on SNP datasets. It includes methods for calculating D-statistics, detecting Fst outliers, computing genetic diversity metrics, AMOVAs, and calculating population genetic summary statistics. This class is particularly useful for researchers studying population structure, diversity, and genetic variation within and between populations. The class is part of the <code class="docutils literal notranslate"><span class="pre">snpio</span></code> package and is designed to work with genotype data loaded using the <code class="docutils literal notranslate"><span class="pre">GenotypeData</span></code> class.</p>
<section id="key-features">
<h2>Key Features<a class="headerlink" href="#key-features" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Comprehensive suite of population genetic analyses for unlinked SNP datasets.</p></li>
<li><p>Calculation of Patterson’s, partitioned, and/ or D-foil D-statistics.</p></li>
<li><p>Fst outlier detection using DBSCAN or bootstrapping. Supports one-tailed and two-tailed P-values.</p></li>
<li><p>Estimation of heterozygosity, nucleotide diversity, and summary statistics.</p></li>
<li><p>Nei’s Distance matrix estimation and visualization between populations with dynamic ordering.</p></li>
<li><p>Hierarchical AMOVA with bootstrapping and parallel bootstrapping computation.</p></li>
</ul>
</section>
<section id="dependencies">
<h2>Dependencies<a class="headerlink" href="#dependencies" title="Link to this heading"></a></h2>
<p>PopGenStatistics is a part of the <code class="docutils literal notranslate"><span class="pre">snpio</span></code> package, which includes classes such as <code class="docutils literal notranslate"><span class="pre">GenotypeEncoder</span></code>, <code class="docutils literal notranslate"><span class="pre">Plotting</span></code>, and modules for calculating D-statistics. It depends on packages such as <code class="docutils literal notranslate"><span class="pre">numpy</span></code>, <code class="docutils literal notranslate"><span class="pre">pandas</span></code>, <code class="docutils literal notranslate"><span class="pre">scipy</span></code>, <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code>, <code class="docutils literal notranslate"><span class="pre">statsmodels</span></code>, <code class="docutils literal notranslate"><span class="pre">seaborn</span></code>, and <code class="docutils literal notranslate"><span class="pre">plotly</span></code>.</p>
</section>
<section id="quick-start">
<h2>Quick Start<a class="headerlink" href="#quick-start" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import classes and initialize GenotypeData with SNP data.</span>
<span class="kn">from</span> <span class="nn">snpio</span> <span class="kn">import</span> <span class="n">VCFReader</span>
<span class="kn">from</span> <span class="nn">snpio.popgenstats</span> <span class="kn">import</span> <span class="n">PopGenStatistics</span>

<span class="c1"># Load SNP data and metadata into a GenotypeData object</span>
<span class="n">genotype_data</span> <span class="o">=</span> <span class="n">VCFReader</span><span class="p">(</span>
    <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;example_data/vcf_files/phylogen_subset14K_sorted.vcf.gz&quot;</span><span class="p">,</span>
    <span class="n">popmapfile</span><span class="o">=</span><span class="s2">&quot;example_data/popmaps/phylogen_nomx.popmap&quot;</span><span class="p">,</span>
    <span class="n">force_popmap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Remove samples not in the popmap or alignment.</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Initialize PopGenStatistics with genotype data</span>
<span class="n">pgs</span> <span class="o">=</span> <span class="n">PopGenStatistics</span><span class="p">(</span><span class="n">genotype_data</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="methods-overview">
<h2>Methods Overview<a class="headerlink" href="#methods-overview" title="Link to this heading"></a></h2>
<table class="responsive-table docutils align-default" id="id1">
<caption><span class="caption-text">Core Statistical Methods</span><a class="headerlink" href="#id1" title="Link to this table"></a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Class Method</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Supported Algorithm(s)</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">calculate_d_statistics</span></code></p></td>
<td><p>Calculates D-statistics and saves them as CSV.</p></td>
<td><p>Patterson’s, partitioned, and D-foil D-statistics.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">detect_fst_outliers</span></code></p></td>
<td><p>Identifies Fst outliers. Supports one-tailed &amp; two-tailed P-values.</p></td>
<td><p>DBSCAN clustering, Traditional bootstrapping.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">summary_statistics</span></code></p></td>
<td><p>Calculates several population genetic summary statistics.</p></td>
<td><p>Observed heterozygosity (Ho), Expected heterozygosity (He), Nucleotide diversity (Pi), Weir and Cockerham’s Fst.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">amova</span></code></p></td>
<td><p>Conducts AMOVA with bootstrapping and parallel computation.</p></td>
<td><p>Hierarchical AMOVA, variance components, Phi statistics.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">neis_genetic_distance</span></code></p></td>
<td><p>Computes Nei’s genetic distance between population pairs.</p></td>
<td><p>Nei’s genetic distance.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="calculated-statistics">
<h2>Calculated Statistics<a class="headerlink" href="#calculated-statistics" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>D-statistics (Patterson’s, Partitioned, and D-foil)</p></li>
<li><p>Fst outliers and pairwise Fst values</p></li>
<li><p>Observed heterozygosity (Ho)</p></li>
<li><p>Expected heterozygosity (He)</p></li>
<li><p>Nucleotide diversity (Pi)</p></li>
<li><p>Weir and Cockerham’s Fst (from summary statistics)</p></li>
<li><p>AMOVA results (variance components and Phi statistics)</p></li>
<li><p>Nei’s genetic distance matrix between populations</p></li>
<li><p>Summary statistics (Ho, He, Pi, Fst)</p></li>
</ul>
</section>
<section id="advanced-usage-and-best-practices">
<h2>Advanced Usage and Best Practices<a class="headerlink" href="#advanced-usage-and-best-practices" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><strong>Distance Matrix Sorting:</strong>
The <code class="docutils literal notranslate"><span class="pre">plot_dist_matrix</span></code> method employs dynamic ordering based on the sum of distances, ensuring that the populations with the highest overall divergence are highlighted.</p></li>
<li><p><strong>Parallelization and Reproducibility:</strong>
Many methods, including <code class="docutils literal notranslate"><span class="pre">amova</span></code>, now support parallel computation via setting <code class="docutils literal notranslate"><span class="pre">n_jobs=-1</span></code> (or any positive integer value), and users are encouraged to define a random seed for consistent results.</p></li>
</ul>
</section>
<section id="additional-information">
<h2>Additional Information<a class="headerlink" href="#additional-information" title="Link to this heading"></a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>Sufficient bootstrap replicates are recommended for robust statistical estimation.</p></li>
<li><p>For parallel processing, set <code class="docutils literal notranslate"><span class="pre">n_jobs=-1</span></code> to utilize all available CPU cores.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">PopGenStatistics</span></code> class is designed for SNP datasets and may not be suitable for other types of genetic data.</p></li>
</ul>
</div>
<p>We hope this documentation helps you take full advantage of the features implemented in the <code class="docutils literal notranslate"><span class="pre">PopGenStatistics</span></code> class. For further details, refer to the source code or the official documentation. If you encounter any issues or have suggestions for improvements, please feel free to reach out to the developers. Thank you for using the SNPio package!</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="example_script.html" class="btn btn-neutral float-left" title="Example script" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="changelog.html" class="btn btn-neutral float-right" title="Changelog" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Bradley T. Martin and Tyler K. Chafin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>