

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>snpio.popgenstats.pop_gen_statistics &mdash; SNPio 1.2.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/custom.css?v=859da57e" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=590429e0"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            SNPio
              <img src="../../../_static/snpio_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../example_script.html">Example script</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pop_gen_statistics.html">PopGenStatistics Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">snpio</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">SNPio</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">snpio.popgenstats.pop_gen_statistics</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for snpio.popgenstats.pop_gen_statistics</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">logging</span> <span class="kn">import</span> <span class="n">Logger</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">multipletests</span>

<span class="kn">from</span> <span class="nn">snpio</span> <span class="kn">import</span> <span class="n">GenotypeEncoder</span><span class="p">,</span> <span class="n">Plotting</span>
<span class="kn">from</span> <span class="nn">snpio.popgenstats.amova</span> <span class="kn">import</span> <span class="n">AMOVA</span>
<span class="kn">from</span> <span class="nn">snpio.popgenstats.d_statistics</span> <span class="kn">import</span> <span class="n">DStatistics</span>
<span class="kn">from</span> <span class="nn">snpio.popgenstats.fst_outliers</span> <span class="kn">import</span> <span class="n">FstOutliers</span>
<span class="kn">from</span> <span class="nn">snpio.popgenstats.genetic_distance</span> <span class="kn">import</span> <span class="n">GeneticDistance</span>
<span class="kn">from</span> <span class="nn">snpio.popgenstats.summary_statistics</span> <span class="kn">import</span> <span class="n">SummaryStatistics</span>
<span class="kn">from</span> <span class="nn">snpio.utils.logging</span> <span class="kn">import</span> <span class="n">LoggerManager</span>
<span class="kn">from</span> <span class="nn">snpio.utils.results_exporter</span> <span class="kn">import</span> <span class="n">ResultsExporter</span>

<span class="n">exporter</span> <span class="o">=</span> <span class="n">ResultsExporter</span><span class="p">(</span><span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;snpio_output&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="PopGenStatistics">
<a class="viewcode-back" href="../../../snpio.popgenstats.html#snpio.popgenstats.pop_gen_statistics.PopGenStatistics">[docs]</a>
<span class="k">class</span> <span class="nc">PopGenStatistics</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for calculating population genetics statistics from SNP data.</span>

<span class="sd">    This class provides methods for calculating population genetics statistics from SNP data. It is designed to work with GenotypeData objects. The PopGenStatistics class can calculate Patterson&#39;s D-statistic, partitioned D-statistic, D-foil statistic, summary statistics, and perform PCA and DAPC dimensionality reduction analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">genotype_data</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">debug</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initialize the PopGenStatistics object.</span>

<span class="sd">        This class provides methods for calculating population genetics statistics from SNP data. It is designed to work with GenotypeData objects. The PopGenStatistics class can calculate Patterson&#39;s D-statistic, partitioned D-statistic, D-foil statistic, summary statistics, and perform PCA and DAPC dimensionality reduction analysis.</span>

<span class="sd">        Args:</span>
<span class="sd">            genotype_data (GenotypeData): GenotypeData object containing SNP data and metadata.</span>
<span class="sd">            verbose (bool): Whether to display verbose output. Defaults to False.</span>
<span class="sd">            debug (bool): Whether to display debug output. Defaults to False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="p">:</span> <span class="n">Any</span> <span class="o">=</span> <span class="n">genotype_data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="n">verbose</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">debug</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="n">debug</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="n">genotype_data</span><span class="o">.</span><span class="n">snp_data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">popmap</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">genotype_data</span><span class="o">.</span><span class="n">popmap</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">populations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">genotype_data</span><span class="o">.</span><span class="n">populations</span>

        <span class="n">plot_kwargs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="n">genotype_data</span><span class="o">.</span><span class="n">plot_kwargs</span>
        <span class="n">plot_kwargs</span><span class="p">[</span><span class="s2">&quot;debug&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">debug</span>
        <span class="n">plot_kwargs</span><span class="p">[</span><span class="s2">&quot;verbose&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">verbose</span>

        <span class="c1"># Initialize plotting and dstats objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plotter</span><span class="p">:</span> <span class="n">Any</span> <span class="o">=</span> <span class="n">Plotting</span><span class="p">(</span><span class="n">genotype_data</span><span class="p">,</span> <span class="o">**</span><span class="n">plot_kwargs</span><span class="p">)</span>

        <span class="c1"># Initialize logger</span>
        <span class="n">logman</span> <span class="o">=</span> <span class="n">LoggerManager</span><span class="p">(</span>
            <span class="vm">__name__</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="n">genotype_data</span><span class="o">.</span><span class="n">prefix</span><span class="p">,</span> <span class="n">debug</span><span class="o">=</span><span class="n">debug</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span>
        <span class="p">)</span>

        <span class="c1"># Get logger object and set logging level</span>
        <span class="n">level</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;DEBUG&quot;</span> <span class="k">if</span> <span class="n">debug</span> <span class="k">else</span> <span class="s2">&quot;INFO&quot;</span>
        <span class="n">logman</span><span class="o">.</span><span class="n">set_level</span><span class="p">(</span><span class="n">level</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="p">:</span> <span class="n">Logger</span> <span class="o">=</span> <span class="n">logman</span><span class="o">.</span><span class="n">get_logger</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">d_stats</span> <span class="o">=</span> <span class="n">DStatistics</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="o">.</span><span class="n">samples</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">logger</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">encoder</span> <span class="o">=</span> <span class="n">GenotypeEncoder</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alignment_012</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">encoder</span><span class="o">.</span><span class="n">genotypes_012</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>

        <span class="n">exporter</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="o">.</span><span class="n">prefix</span><span class="si">}</span><span class="s2">_output&quot;</span><span class="p">,</span> <span class="s2">&quot;analysis&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="PopGenStatistics.calculate_d_statistics">
<a class="viewcode-back" href="../../../snpio.popgenstats.html#snpio.popgenstats.pop_gen_statistics.PopGenStatistics.calculate_d_statistics">[docs]</a>
    <span class="nd">@exporter</span><span class="o">.</span><span class="n">capture_results</span>
    <span class="k">def</span> <span class="nf">calculate_d_statistics</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">population1</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">population2</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">population3</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">outgroup</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">population4</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">include_heterozygous</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">num_bootstraps</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
        <span class="n">n_jobs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">max_individuals_per_pop</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">individual_selection</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;random&quot;</span><span class="p">,</span>
        <span class="n">output_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">save_plot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate D-statistics, Z-scores, and P-values from D-statistics.</span>

<span class="sd">        This method calculates D-statistics, Z-scores, and P-values for all possible sample combinations based on the chosen ``method``. The results are saved to a CSV file and returned as a pandas DataFrame. The overall mean D-statistic, Z-score, and P-value are also returned as a dictionary.</span>

<span class="sd">        Args:</span>
<span class="sd">            method (str): The method to use for calculating D-statistics. Supported options include &quot;patterson&quot;, &quot;partitioned&quot;, &quot;dfoil&quot;. &quot;patterson&quot; is the default and represents the traditional 4-taxon D-statistic. &quot;partitioned&quot; is a generalization of the D-statistic that allows for more than 4 populations. &quot;dfoil&quot; is a method that calculates D-statistics for all possible combinations of 4 populations (i.e., the FOIL method). Defaults to &quot;patterson&quot;.</span>
<span class="sd">            population1 (str | List[str]): Population ID or list of IDs.</span>
<span class="sd">            population2 (str | List[str]): Population ID or list of IDs.</span>
<span class="sd">            population3 (str | List[str]): Population ID or list of IDs.</span>
<span class="sd">            outgroup (str | List[str]): Population ID or list of IDs.</span>
<span class="sd">            population4 (str | List[str], optional): Population ID or list of IDs. Required for &quot;partitioned&quot; and &quot;dfoil&quot; methods.</span>
<span class="sd">            include_heterozygous (bool): Whether to include heterozygous genotypes. Defaults to False.</span>
<span class="sd">            num_bootstraps (int): Number of bootstrap replicates.</span>
<span class="sd">            n_jobs (int): Number of parallel jobs. -1 uses all available cores. Defaults to -1.</span>
<span class="sd">            max_individuals_per_pop (Optional[int]): Max individuals per population. Defaults to None. If specified, will select individuals from each population based on the criteria specified in ``individual_selection``.</span>
<span class="sd">            individual_selection (str | Dict[str, List[str]]): Method for selecting individuals. Defaults to &quot;random&quot;. If max_individuals_per_pop is specified, can be a dictionary with population IDs as keys and lists of selected individuals as values.</span>
<span class="sd">            output_file (Optional[str]): Path to save the results CSV file. If not specified, results will be saved to a default location. Defaults to None.</span>
<span class="sd">            save_plot (bool): Whether to save the plots D-statistic plots. Defaults to True.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple: A tuple containing the results of all sample combinations as a pandas DataFrame and the overall mean D-statistic, Z-score, and P-value as a dictionary.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Calculating </span><span class="si">{</span><span class="n">method</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s2"> D-statistics...&quot;</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">get_population_indices</span><span class="p">(</span><span class="n">population</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;Retrieve sample indices for a population or list of populations.</span>

<span class="sd">            Args:</span>
<span class="sd">                population (Union[str, List[str]]): Population ID or list of IDs.</span>

<span class="sd">            Returns:</span>
<span class="sd">                List[int]: List of sample indices.</span>

<span class="sd">            Raises:</span>
<span class="sd">                ValueError: If a population ID is not found.</span>
<span class="sd">                ValueError: If an invalid ``individual_selection`` method is specified.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">populations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="p">[</span><span class="n">population</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">population</span>
            <span class="p">)</span>
            <span class="n">selected_samples</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">popmap_inverse</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="o">.</span><span class="n">popmap_inverse</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">samples</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">popmap_inverse</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span>
                    <span class="c1"># Limit individuals per population if specified</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">max_individuals_per_pop</span>
                        <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_individuals_per_pop</span>
                    <span class="p">):</span>
                        <span class="k">if</span> <span class="n">individual_selection</span> <span class="o">==</span> <span class="s2">&quot;random&quot;</span><span class="p">:</span>
                            <span class="n">selected_samples</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                                <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
                                    <span class="n">samples</span><span class="p">,</span> <span class="n">max_individuals_per_pop</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span>
                                <span class="p">)</span>
                            <span class="p">)</span>
                        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">individual_selection</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
                            <span class="n">selected_samples</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                                <span class="n">individual_selection</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                                    <span class="n">pop</span><span class="p">,</span> <span class="n">samples</span><span class="p">[:</span><span class="n">max_individuals_per_pop</span><span class="p">]</span>
                                <span class="p">)</span>
                            <span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">msg</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="p">(</span>
                                <span class="sa">f</span><span class="s2">&quot;Invalid individual_selection: &#39;</span><span class="si">{</span><span class="n">individual_selection</span><span class="si">}</span><span class="s2">&#39;.&quot;</span>
                            <span class="p">)</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
                            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">selected_samples</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Population ID &#39;</span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s2">&#39; not found.&quot;</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Population ID &#39;</span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s2">&#39; not found.&quot;</span><span class="p">)</span>

            <span class="c1"># Convert sample IDs to indices</span>
            <span class="n">samples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="o">.</span><span class="n">samples</span>
            <span class="n">sample_id_to_index</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">sample</span><span class="p">:</span> <span class="n">idx</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="k">return</span> <span class="p">[</span>
                <span class="n">sample_id_to_index</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">selected_samples</span>
                <span class="k">if</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sample_id_to_index</span>
            <span class="p">]</span>

        <span class="c1"># Retrieve sample indices for each population</span>
        <span class="n">d1_inds</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_population_indices</span><span class="p">(</span><span class="n">population1</span><span class="p">)</span>
        <span class="n">d2_inds</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_population_indices</span><span class="p">(</span><span class="n">population2</span><span class="p">)</span>
        <span class="n">d3_inds</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_population_indices</span><span class="p">(</span><span class="n">population3</span><span class="p">)</span>
        <span class="n">d4_inds</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">get_population_indices</span><span class="p">(</span><span class="n">population4</span><span class="p">)</span> <span class="k">if</span> <span class="n">population4</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="p">)</span>
        <span class="n">outgroup_inds</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_population_indices</span><span class="p">(</span><span class="n">outgroup</span><span class="p">)</span>

        <span class="c1"># Calculate Z-scores and P-values</span>
        <span class="n">combo_z_p_values</span><span class="p">,</span> <span class="n">overall_z_score</span><span class="p">,</span> <span class="n">overall_p_value</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">d_stats</span><span class="o">.</span><span class="n">calculate_z_and_p_values</span><span class="p">(</span>
                <span class="n">d1_inds</span><span class="p">,</span>
                <span class="n">d2_inds</span><span class="p">,</span>
                <span class="n">d3_inds</span><span class="p">,</span>
                <span class="n">outgroup_inds</span><span class="p">,</span>
                <span class="n">d4_inds</span><span class="p">,</span>
                <span class="n">include_heterozygous</span><span class="p">,</span>
                <span class="n">num_bootstraps</span><span class="p">,</span>
                <span class="n">method</span><span class="p">,</span>
                <span class="n">n_jobs</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Directory and filename setup</span>
        <span class="k">if</span> <span class="n">output_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">outdir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="o">.</span><span class="n">prefix</span><span class="si">}</span><span class="s2">_output&quot;</span><span class="p">,</span> <span class="s2">&quot;analysis&quot;</span><span class="p">,</span> <span class="s2">&quot;d_stats&quot;</span><span class="p">)</span>
            <span class="n">outdir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">output_file</span> <span class="o">=</span> <span class="n">outdir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_dstat_results.csv&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
            <span class="n">output_file</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Prepare columns for sample combinations</span>
        <span class="n">max_combo_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">combo</span><span class="p">)</span> <span class="k">for</span> <span class="n">combo</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">combo_z_p_values</span><span class="p">)</span>
        <span class="n">combo_columns</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Sample_</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_combo_len</span><span class="p">)]</span>

        <span class="c1"># Create data entries with columns for each sample in combination</span>
        <span class="n">data</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">combo</span><span class="p">,</span> <span class="n">observed_d</span><span class="p">,</span> <span class="n">z</span><span class="p">,</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">combo_z_p_values</span><span class="p">:</span>
            <span class="n">combo_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">combo_columns</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span> <span class="nb">str</span><span class="p">(</span><span class="n">combo</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">combo</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_combo_len</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="n">combo_data</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;Observed D-Statistic&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">observed_d</span><span class="p">),</span>
                    <span class="s2">&quot;Z-Score&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">z</span><span class="p">),</span>
                    <span class="s2">&quot;P-Value&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">p</span><span class="p">),</span>
                <span class="p">}</span>
            <span class="p">)</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">combo_data</span><span class="p">)</span>

        <span class="c1"># Get overall mean results.</span>
        <span class="n">overall_d_stat</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">floating</span><span class="p">[</span><span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span>
            <span class="p">[</span><span class="n">obs</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">obs</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">combo_z_p_values</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">overall_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Observed D-Statistic&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">overall_d_stat</span><span class="p">),</span>
            <span class="s2">&quot;Z-Score&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">overall_z_score</span><span class="p">),</span>
            <span class="s2">&quot;P-Value&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">overall_p_value</span><span class="p">),</span>
        <span class="p">}</span>

        <span class="c1"># Create DataFrame from data</span>
        <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="c1"># Bonferroni and FDR corrections</span>
        <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_adjust_p_values</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="c1"># Save results to CSV</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Results saved to </span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">save_plot</span><span class="p">:</span>
            <span class="c1"># Plot D-statistic results</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plotter</span><span class="o">.</span><span class="n">plot_d_statistics</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;D-statistics calculation complete!&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span><span class="p">,</span> <span class="n">overall_data</span></div>


    <span class="k">def</span> <span class="nf">_adjust_p_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adjust P-values using multiple testing correction methods.</span>

<span class="sd">        This method adjusts P-values using the Bonferroni and FDR B-H methods. The adjusted P-values are added to the DataFrame, along with significance columns for each correction.</span>

<span class="sd">        Args:</span>
<span class="sd">            df (pd.DataFrame): DataFrame containing P-values.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: DataFrame with adjusted P-values and significance columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Adjusting P-values...&quot;</span><span class="p">)</span>

        <span class="c1"># Make a copy of the DataFrame to avoid modifying the original</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Bonferroni P-Value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">multipletests</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;P-Value&quot;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;bonferroni&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;FDR-BH P-Value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">multipletests</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;P-Value&quot;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;fdr_bh&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1"># Add significance columns for each correction</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Significant (Raw)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;P-Value&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Significant (Bonferroni)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Bonferroni P-Value&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Significant (FDR-BH)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;FDR-BH P-Value&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span>

        <span class="k">return</span> <span class="n">df</span>

<div class="viewcode-block" id="PopGenStatistics.detect_fst_outliers">
<a class="viewcode-back" href="../../../snpio.popgenstats.html#snpio.popgenstats.pop_gen_statistics.PopGenStatistics.detect_fst_outliers">[docs]</a>
    <span class="k">def</span> <span class="nf">detect_fst_outliers</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">correction_method</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;bonf&quot;</span><span class="p">,</span> <span class="s2">&quot;fdr&quot;</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">alpha</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
        <span class="n">use_bootstrap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">n_bootstraps</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
        <span class="n">n_jobs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">tail_direction</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="s2">&quot;upper&quot;</span><span class="p">,</span> <span class="s2">&quot;lower&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Detect Fst outliers from SNP data using bootstrapping or DBSCAN.</span>

<span class="sd">        This method detects Fst outliers from SNP data using bootstrapping or DBSCAN clustering. Outliers are identified based on the distribution of Fst values between population pairs. The method returns a DataFrame containing the Fst outliers and contributing population pairs, as well as a DataFrame containing the adjusted or unadjusted P-values, depending on whether a multiple testing correction method was specified.</span>

<span class="sd">        Args:</span>
<span class="sd">            correction_method (str, optional): Multiple testing correction method that performs P-value adjustment, &#39;bonf&#39; (Bonferroni) or &#39;fdr&#39; (FDR B-H). If not specified, no correction or P-value adjustment is applied. Defaults to None.</span>
<span class="sd">            alpha (float): Significance level for multiple test correction (with adjusted P-values). Defaults to 0.05.</span>
<span class="sd">            use_bootstrap (bool): Whether to use bootstrapping to estimate variance of Fst per SNP. If False, DBSCAN clustering is used instead. Defaults to False.</span>
<span class="sd">            n_bootstraps (int): Number of bootstrap replicates to use for estimating variance of Fst per SNP. Defaults to 1000.</span>
<span class="sd">            n_jobs (int): Number of CPU threads to use for parallelization. If set to -1, all available CPU threads are used. Defaults to 1.</span>
<span class="sd">            tail_direction (str): Direction of the test for Fst outliers. &quot;both&quot; for two-tailed, &quot;upper&quot; for testing higher than expected, and &quot;lower&quot; for lower than expected. Defaults to &quot;both&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple[pd.DataFrame, pd.DataFrame]: A DataFrame containing the Fst outliers and contributing population pairs, and a DataFrame containing the adjusted P-values if ``correction_method`` was provided or the un-adjusted P-values otherwise.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Detecting Fst outliers...&quot;</span><span class="p">)</span>

        <span class="n">fo</span> <span class="o">=</span> <span class="n">FstOutliers</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignment_012</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">plotter</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">correction_method</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">correction_method</span> <span class="o">=</span> <span class="n">correction_method</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">correction_method</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;bonf&quot;</span><span class="p">,</span> <span class="s2">&quot;fdr&quot;</span><span class="p">}:</span>
                <span class="n">msg</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Invalid correction_method. Supported options: &#39;bonf&#39;, &#39;fdr&#39;, but got: </span><span class="si">{</span><span class="n">correction_method</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

            <span class="n">correction_method</span> <span class="o">=</span> <span class="s2">&quot;fdr_bh&quot;</span> <span class="k">if</span> <span class="n">correction_method</span> <span class="o">==</span> <span class="s2">&quot;fdr&quot;</span> <span class="k">else</span> <span class="s2">&quot;bonferroni&quot;</span>

        <span class="k">if</span> <span class="n">alpha</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">alpha</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">msg</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Invalid alpha value. Must be in the range [0, 1), but got: </span><span class="si">{</span><span class="n">alpha</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">n_bootstraps</span> <span class="o">&lt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">n_bootstraps</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">msg</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Invalid n_bootstraps value. Must be an integer greater than 0, but got: </span><span class="si">{</span><span class="n">n_bootstraps</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="n">tail_direction</span> <span class="o">=</span> <span class="n">tail_direction</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">tail_direction</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="s2">&quot;upper&quot;</span><span class="p">,</span> <span class="s2">&quot;lower&quot;</span><span class="p">}:</span>
            <span class="n">msg</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Invalid tail_direction. Supported options: &#39;both&#39;, &#39;upper&#39;, &#39;lower&#39;, but got: </span><span class="si">{</span><span class="n">tail_direction</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="n">alternative</span> <span class="o">=</span> <span class="s2">&quot;two&quot;</span> <span class="k">if</span> <span class="n">tail_direction</span> <span class="o">==</span> <span class="s2">&quot;both&quot;</span> <span class="k">else</span> <span class="n">tail_direction</span>

        <span class="c1"># Returns tuple of outlier SNPs and adjusted p-values.</span>
        <span class="k">if</span> <span class="n">use_bootstrap</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">fo</span><span class="o">.</span><span class="n">detect_fst_outliers_bootstrap</span><span class="p">(</span>
                <span class="n">correction_method</span><span class="p">,</span>
                <span class="n">alpha</span><span class="p">,</span>
                <span class="n">n_bootstraps</span><span class="p">,</span>
                <span class="n">n_jobs</span><span class="p">,</span>
                <span class="n">alternative</span><span class="o">=</span><span class="n">alternative</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">fo</span><span class="o">.</span><span class="n">detect_fst_outliers_dbscan</span><span class="p">(</span><span class="n">correction_method</span><span class="p">,</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">n_jobs</span><span class="p">)</span></div>


    <span class="c1"># @exporter.capture_results</span>
<div class="viewcode-block" id="PopGenStatistics.summary_statistics">
<a class="viewcode-back" href="../../../snpio.popgenstats.html#snpio.popgenstats.pop_gen_statistics.PopGenStatistics.summary_statistics">[docs]</a>
    <span class="k">def</span> <span class="nf">summary_statistics</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">n_bootstraps</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">save_plots</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">use_pvalues</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate a suite of summary statistics for SNP data.</span>

<span class="sd">        This method calculates a suite of summary statistics for SNP data, including observed heterozygosity (Ho), expected heterozygosity (He), nucleotide diversity (Pi), and Fst between populations. Summary statistics are calculated both overall and per population.</span>

<span class="sd">        Args:</span>
<span class="sd">            n_bootstraps (int): Number of bootstrap replicates to use for estimating variance of Fst per SNP. If 0, then bootstrapping is not used and confidence intervals are estimated from the data. Defaults to 0.</span>
<span class="sd">            n_jobs (int): Number of parallel jobs. If set to -1, all available CPU threads are used. Defaults to 1.</span>
<span class="sd">            save_plots (bool): Whether to save plots of the summary statistics. In any case, a dictionary of summary statistics is returned. Defaults to True.</span>
<span class="sd">            use_pvalues (bool): Whether to calculate p-values for Fst. Otherwise calculates 95% confidence intervals. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary containing summary statistics per population and overall.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">summary_stats</span> <span class="o">=</span> <span class="n">SummaryStatistics</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">alignment_012</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plotter</span><span class="p">,</span>
            <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">,</span>
            <span class="n">debug</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">debug</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">summary_stats</span><span class="o">.</span><span class="n">calculate_summary_statistics</span><span class="p">(</span>
            <span class="n">n_bootstraps</span><span class="o">=</span><span class="n">n_bootstraps</span><span class="p">,</span>
            <span class="n">n_jobs</span><span class="o">=</span><span class="n">n_jobs</span><span class="p">,</span>
            <span class="n">save_plots</span><span class="o">=</span><span class="n">save_plots</span><span class="p">,</span>
            <span class="n">use_pvalues</span><span class="o">=</span><span class="n">use_pvalues</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="PopGenStatistics.amova">
<a class="viewcode-back" href="../../../snpio.popgenstats.html#snpio.popgenstats.pop_gen_statistics.PopGenStatistics.amova">[docs]</a>
    <span class="nd">@exporter</span><span class="o">.</span><span class="n">capture_results</span>
    <span class="k">def</span> <span class="nf">amova</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">regionmap</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">n_bootstraps</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">n_jobs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">random_seed</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Conduct AMOVA (Analysis Of Molecular Variance).</span>

<span class="sd">        AMOVA (Analysis of Molecular Variance) is a method for partitioning genetic variation into components due to differences among populations, among individuals within populations, and within individuals. This method calculates variance components and Phi statistics for a given number of hierarchical levels (1, 2, or 3). If bootstrapping is enabled, it also estimates p-values for the variance components. The number of hierarchical levels determines the structure of the AMOVA model: 1 =&gt; populations only, 2 =&gt; region -&gt; populations, 3 =&gt; region -&gt; population -&gt; individuals. If regionmap is provided, it is used to map populations to regions.</span>

<span class="sd">        Notes:</span>
<span class="sd">            - Algorithm adapted from the R package &#39;poppr&#39; (Kamvar et al., 2014).</span>
<span class="sd">            - The Phi statistic is a measure of genetic differentiation between populations.</span>
<span class="sd">            - Algorithm description: First, the total variance is calculated. Then, the variance components are calculated by summing the squared differences between the mean of each group and the global mean. The Phi statistic is calculated as the ratio of the among-group variance to the total variance. P-values are estimated by bootstrapping.</span>

<span class="sd">        Args:</span>
<span class="sd">            regionmap (dict, optional): Mapping from population_id -&gt; region_id.</span>
<span class="sd">                If None but hierarchical_levels&gt;1, raises ValueError.</span>
<span class="sd">            n_bootstraps (int): Number of bootstrap replicates across SNP loci.</span>
<span class="sd">            n_jobs (int): Number of parallel jobs. -1 uses all cores.</span>
<span class="sd">            random_seed (int, optional): Random seed for reproducibility.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: AMOVA results (variance components, Phi statistics, and possibly p-values).</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If regionmap is required but not provided.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">amova_instance</span> <span class="o">=</span> <span class="n">AMOVA</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">amova_instance</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
            <span class="n">regionmap</span><span class="o">=</span><span class="n">regionmap</span><span class="p">,</span>
            <span class="n">n_bootstraps</span><span class="o">=</span><span class="n">n_bootstraps</span><span class="p">,</span>
            <span class="n">n_jobs</span><span class="o">=</span><span class="n">n_jobs</span><span class="p">,</span>
            <span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="PopGenStatistics.neis_genetic_distance">
<a class="viewcode-back" href="../../../snpio.popgenstats.html#snpio.popgenstats.pop_gen_statistics.PopGenStatistics.neis_genetic_distance">[docs]</a>
    <span class="nd">@exporter</span><span class="o">.</span><span class="n">capture_results</span>
    <span class="k">def</span> <span class="nf">neis_genetic_distance</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">n_bootstraps</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">n_jobs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">use_pvalues</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">palette</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;magma&quot;</span><span class="p">,</span>
        <span class="n">supress_plot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">|</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate Nei&#39;s genetic distance between all pairs of populations.</span>

<span class="sd">        Optionally computes bootstrap-based p-values for each population pair if n_bootstraps &gt; 0.</span>

<span class="sd">        Nei&#39;s genetic distance is defined as ``D = -ln(  )``, where  is the ratio of the average genetic identity to the geometric mean of the average homozygosities.</span>

<span class="sd">        Args:</span>
<span class="sd">            n_bootstraps (int): Number of bootstrap replicates to compute p-values. Defaults to 0 (only distances are returned).</span>
<span class="sd">            n_jobs (int): Number of parallel jobs. -1 uses all cores. Defaults to 1.</span>
<span class="sd">            use_pvalues (bool): If True, returns a tuple of (distance matrix, p-value matrix). Defaults to False.</span>
<span class="sd">            palette (str): Color palette for the distance matrix plot. Can use any matplotlib gradient-based palette. Some frequently used options include: &quot;coolwarm&quot;, &quot;viridis&quot;, &quot;magma&quot;, and &quot;inferno&quot;. Defaults to &#39;coolwarm&#39;.</span>
<span class="sd">            supress_plot (bool): If True, suppresses the plotting of the distance matrix. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: If n_bootstraps == 0, returns a DataFrame of Nei&#39;s distances.</span>
<span class="sd">            Tuple[pd.DataFrame, pd.DataFrame]: If n_bootstraps &gt; 0, returns a tuple of (distance matrix, p-value matrix).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gd</span> <span class="o">=</span> <span class="n">GeneticDistance</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">plotter</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">,</span> <span class="n">debug</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">debug</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating Nei&#39;s genetic distance...&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of bootstraps: </span><span class="si">{</span><span class="n">n_bootstraps</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">nei_results</span> <span class="o">=</span> <span class="n">gd</span><span class="o">.</span><span class="n">nei_distance</span><span class="p">(</span>
            <span class="n">n_bootstraps</span><span class="o">=</span><span class="n">n_bootstraps</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="n">n_jobs</span><span class="p">,</span> <span class="n">return_pvalues</span><span class="o">=</span><span class="n">use_pvalues</span>
        <span class="p">)</span>

        <span class="n">df_obs</span><span class="p">,</span> <span class="n">df_lower</span><span class="p">,</span> <span class="n">df_upper</span><span class="p">,</span> <span class="n">df_pval</span> <span class="o">=</span> <span class="n">gd</span><span class="o">.</span><span class="n">parse_nei_result</span><span class="p">(</span><span class="n">nei_results</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">supress_plot</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plotter</span><span class="o">.</span><span class="n">plot_dist_matrix</span><span class="p">(</span>
                <span class="n">df_obs</span><span class="p">,</span>
                <span class="n">pvals</span><span class="o">=</span><span class="n">df_pval</span> <span class="k">if</span> <span class="n">use_pvalues</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>
                <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Nei&#39;s Genetic Distance&quot;</span><span class="p">,</span>
                <span class="n">dist_type</span><span class="o">=</span><span class="s2">&quot;nei&quot;</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Nei&#39;s genetic distance calculation complete!&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">df_obs</span><span class="p">,</span> <span class="n">df_pval</span><span class="p">)</span> <span class="k">if</span> <span class="n">use_pvalues</span> <span class="k">else</span> <span class="n">df_obs</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Bradley T. Martin and Tyler K. Chafin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>