

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Example script &mdash; SNPio 1.2.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=859da57e" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=590429e0"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="PopGenStatistics Documentation" href="pop_gen_statistics.html" />
    <link rel="prev" title="Getting Started" href="getting_started.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SNPio
              <img src="_static/snpio_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Example script</a></li>
<li class="toctree-l1"><a class="reference internal" href="pop_gen_statistics.html">PopGenStatistics Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">snpio</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SNPio</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Example script</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/btmartin721/SNPio/blob/master/docs/source/example_script.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="example-script">
<h1>Example script<a class="headerlink" href="#example-script" title="Link to this heading">ÔÉÅ</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">run_snpio.py</span></code> script provides a template you can use to get started. It demonstrates how to use the main classes in the package. The script reads a VCF file, a popmap file, and a tree file, and then runs various analyses on the data. The script also demonstrates how to encode genotypes into different formats, filter the data, and parse a tree file. The script is located in the <code class="docutils literal notranslate"><span class="pre">snpio</span></code> directory. To run the script, navigate to the <code class="docutils literal notranslate"><span class="pre">snpio</span></code> directory and run the following command:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>python3<span class="w"> </span>run_snpio.py
</pre></div>
</div>
<p>It will then run the example data.</p>
<p>Below is the code for the script:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">snpio</span> <span class="kn">import</span> <span class="p">(</span>
   <span class="n">GenotypeEncoder</span><span class="p">,</span>
   <span class="n">NRemover2</span><span class="p">,</span>
   <span class="n">Plotting</span><span class="p">,</span>
   <span class="n">PopGenStatistics</span><span class="p">,</span>
   <span class="n">TreeParser</span><span class="p">,</span>
   <span class="n">VCFReader</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Uncomment the following line to import the Benchmark class.</span>
<span class="c1"># NOTE: For development purposes. Comment out for normal use.</span>
<span class="c1"># from snpio.utils.benchmarking import Benchmark</span>


<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
   <span class="c1"># Read the alignment, popmap, and tree files.</span>
   <span class="n">gd</span> <span class="o">=</span> <span class="n">VCFReader</span><span class="p">(</span>
      <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;snpio/example_data/vcf_files/phylogen_subset14K_sorted.vcf.gz&quot;</span><span class="p">,</span>
      <span class="n">popmapfile</span><span class="o">=</span><span class="s2">&quot;snpio/example_data/popmaps/phylogen_nomx.popmap&quot;</span><span class="p">,</span>
      <span class="n">force_popmap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Remove samples not in the popmap, or vice versa.</span>
      <span class="n">chunk_size</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
      <span class="n">exclude_pops</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;OG&quot;</span><span class="p">],</span>
      <span class="n">plot_format</span><span class="o">=</span><span class="s2">&quot;pdf&quot;</span><span class="p">,</span>
   <span class="p">)</span>

   <span class="n">pgs</span> <span class="o">=</span> <span class="n">PopGenStatistics</span><span class="p">(</span><span class="n">gd</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

   <span class="n">summary_stats</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">summary_statistics</span><span class="p">(</span><span class="n">save_plots</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

   <span class="n">df_fst_outliers_boot</span><span class="p">,</span> <span class="n">df_fst_outlier_pvalues_boot</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">detect_fst_outliers</span><span class="p">(</span>
      <span class="n">correction_method</span><span class="o">=</span><span class="s2">&quot;bonf&quot;</span><span class="p">,</span>
      <span class="n">use_bootstrap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
      <span class="n">n_bootstraps</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
      <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
      <span class="n">tail_direction</span><span class="o">=</span><span class="s2">&quot;upper&quot;</span><span class="p">,</span>
   <span class="p">)</span>

   <span class="n">df_fst_outliers_dbscan</span><span class="p">,</span> <span class="n">df_fst_outlier_pvalues_dbscan</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">detect_fst_outliers</span><span class="p">(</span>
      <span class="n">correction_method</span><span class="o">=</span><span class="s2">&quot;bonf&quot;</span><span class="p">,</span> <span class="n">use_bootstrap</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span>
   <span class="p">)</span>

   <span class="c1"># NOTE: Takes a while to run.</span>
   <span class="n">amova_results</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">amova</span><span class="p">(</span>
      <span class="n">regionmap</span><span class="o">=</span><span class="p">{</span>
         <span class="s2">&quot;EA&quot;</span><span class="p">:</span> <span class="s2">&quot;Eastern&quot;</span><span class="p">,</span>
         <span class="s2">&quot;GU&quot;</span><span class="p">:</span> <span class="s2">&quot;Eastern&quot;</span><span class="p">,</span>
         <span class="s2">&quot;TT&quot;</span><span class="p">:</span> <span class="s2">&quot;Eastern&quot;</span><span class="p">,</span>
         <span class="s2">&quot;TC&quot;</span><span class="p">:</span> <span class="s2">&quot;Eastern&quot;</span><span class="p">,</span>
            <span class="s2">&quot;DS&quot;</span><span class="p">:</span> <span class="s2">&quot;Ornate&quot;</span><span class="p">,</span>
      <span class="p">},</span>
      <span class="n">n_bootstraps</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
      <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
      <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
   <span class="p">)</span>

   <span class="n">nei_dist_df</span><span class="p">,</span> <span class="n">nei_pvals_df</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">neis_genetic_distance</span><span class="p">(</span><span class="n">n_bootstraps</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>

   <span class="n">dstats_df</span><span class="p">,</span> <span class="n">overall_results</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">calculate_d_statistics</span><span class="p">(</span>
      <span class="n">method</span><span class="o">=</span><span class="s2">&quot;patterson&quot;</span><span class="p">,</span>
      <span class="n">population1</span><span class="o">=</span><span class="s2">&quot;EA&quot;</span><span class="p">,</span>
      <span class="n">population2</span><span class="o">=</span><span class="s2">&quot;GU&quot;</span><span class="p">,</span>
      <span class="n">population3</span><span class="o">=</span><span class="s2">&quot;TT&quot;</span><span class="p">,</span>
      <span class="n">outgroup</span><span class="o">=</span><span class="s2">&quot;ON&quot;</span><span class="p">,</span>
      <span class="n">num_bootstraps</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
      <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
      <span class="n">max_individuals_per_pop</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
   <span class="p">)</span>

   <span class="c1"># # Run PCA and make missingness report plots.</span>
   <span class="n">plotting</span> <span class="o">=</span> <span class="n">Plotting</span><span class="p">(</span><span class="n">genotype_data</span><span class="o">=</span><span class="n">gd</span><span class="p">)</span>
   <span class="n">gd_components</span><span class="p">,</span> <span class="n">gd_pca</span> <span class="o">=</span> <span class="n">plotting</span><span class="o">.</span><span class="n">run_pca</span><span class="p">()</span>
   <span class="n">gd</span><span class="o">.</span><span class="n">missingness_reports</span><span class="p">()</span>

   <span class="n">nrm</span> <span class="o">=</span> <span class="n">NRemover2</span><span class="p">(</span><span class="n">gd</span><span class="p">)</span>

   <span class="n">nrm</span><span class="o">.</span><span class="n">search_thresholds</span><span class="p">(</span>
      <span class="n">thresholds</span><span class="o">=</span><span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">],</span>
      <span class="n">maf_thresholds</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">],</span>
      <span class="n">mac_thresholds</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
   <span class="p">)</span>

   <span class="c1"># Plot benchmarking results.</span>
   <span class="c1"># NOTE: For development purposes. Comment out for normal use.</span>
   <span class="c1"># Benchmark.plot_performance(nrm.genotype_data, nrm.genotype_data.resource_data)</span>

   <span class="n">gd_filt</span> <span class="o">=</span> <span class="p">(</span>
      <span class="n">nrm</span><span class="o">.</span><span class="n">filter_missing_sample</span><span class="p">(</span><span class="mf">0.75</span><span class="p">)</span>
      <span class="o">.</span><span class="n">filter_missing</span><span class="p">(</span><span class="mf">0.75</span><span class="p">)</span>
      <span class="o">.</span><span class="n">filter_missing_pop</span><span class="p">(</span><span class="mf">0.75</span><span class="p">)</span>
      <span class="o">.</span><span class="n">filter_mac</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
      <span class="o">.</span><span class="n">filter_monomorphic</span><span class="p">(</span><span class="n">exclude_heterozygous</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
      <span class="o">.</span><span class="n">filter_singletons</span><span class="p">(</span><span class="n">exclude_heterozygous</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
      <span class="o">.</span><span class="n">filter_biallelic</span><span class="p">(</span><span class="n">exclude_heterozygous</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
      <span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
   <span class="p">)</span>

   <span class="n">nrm</span><span class="o">.</span><span class="n">plot_sankey_filtering_report</span><span class="p">()</span>

   <span class="c1"># # Make missingness report plots.</span>
   <span class="n">plotting2</span> <span class="o">=</span> <span class="n">Plotting</span><span class="p">(</span><span class="n">genotype_data</span><span class="o">=</span><span class="n">gd_filt</span><span class="p">)</span>
   <span class="n">filt_components</span><span class="p">,</span> <span class="n">filt_pca</span> <span class="o">=</span> <span class="n">plotting2</span><span class="o">.</span><span class="n">run_pca</span><span class="p">()</span>
   <span class="n">gd_filt</span><span class="o">.</span><span class="n">missingness_reports</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;filtered&quot;</span><span class="p">)</span>

   <span class="c1"># # Write the filtered VCF file.</span>
   <span class="n">gd_filt</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="s2">&quot;snpio/example_data/vcf_files/nremover_test.vcf&quot;</span><span class="p">)</span>

   <span class="c1"># # Encode the genotypes into 012, one-hot, and integer formats.</span>
   <span class="n">ge</span> <span class="o">=</span> <span class="n">GenotypeEncoder</span><span class="p">(</span><span class="n">gd_filt</span><span class="p">)</span>
   <span class="n">gt_012</span> <span class="o">=</span> <span class="n">ge</span><span class="o">.</span><span class="n">genotypes_012</span>
   <span class="n">gt_onehot</span> <span class="o">=</span> <span class="n">ge</span><span class="o">.</span><span class="n">genotypes_onehot</span>
   <span class="n">gt_int</span> <span class="o">=</span> <span class="n">ge</span><span class="o">.</span><span class="n">genotypes_int</span>

   <span class="n">df012</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gt_012</span><span class="p">)</span>
   <span class="n">dfint</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gt_int</span><span class="p">)</span>

   <span class="n">tp</span> <span class="o">=</span> <span class="n">TreeParser</span><span class="p">(</span>
      <span class="n">genotype_data</span><span class="o">=</span><span class="n">gd_filt</span><span class="p">,</span>
      <span class="n">treefile</span><span class="o">=</span><span class="s2">&quot;snpio/example_data/trees/test.tre&quot;</span><span class="p">,</span>
      <span class="n">qmatrix</span><span class="o">=</span><span class="s2">&quot;snpio/example_data/trees/test.iqtree&quot;</span><span class="p">,</span>
      <span class="n">siterates</span><span class="o">=</span><span class="s2">&quot;snpio/example_data/trees/test14K.rate&quot;</span><span class="p">,</span>
      <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
      <span class="n">debug</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
   <span class="p">)</span>

   <span class="c1"># # Get a toytree object by reading the tree file.</span>
   <span class="n">tree</span> <span class="o">=</span> <span class="n">tp</span><span class="o">.</span><span class="n">read_tree</span><span class="p">()</span>

   <span class="c1"># # Reroot the tree at any nodes containing the string &#39;EA&#39; in the sampleID.</span>
   <span class="n">tp</span><span class="o">.</span><span class="n">reroot_tree</span><span class="p">(</span><span class="s2">&quot;~EA&quot;</span><span class="p">)</span>

   <span class="c1"># # Get a subtree with only the samples containing &#39;EA&#39; in the sampleID.</span>
   <span class="n">subtree</span> <span class="o">=</span> <span class="n">tp</span><span class="o">.</span><span class="n">get_subtree</span><span class="p">(</span><span class="s2">&quot;~EA&quot;</span><span class="p">)</span>

   <span class="c1"># # Prune the tree to remove samples containing &#39;ON&#39; in the sampleID.</span>
   <span class="n">pruned_tree</span> <span class="o">=</span> <span class="n">tp</span><span class="o">.</span><span class="n">prune_tree</span><span class="p">(</span><span class="s2">&quot;~ON&quot;</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
   <span class="n">main</span><span class="p">()</span>
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="getting_started.html" class="btn btn-neutral float-left" title="Getting Started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pop_gen_statistics.html" class="btn btn-neutral float-right" title="PopGenStatistics Documentation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Bradley T. Martin and Tyler K. Chafin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>